### PR TITLE
Gif command Tenor api

### PR DESCRIPTION
**Pull request by Cor-Lapidem**
The gif command will now use the tenor api and choose the first gif it returns, if the keyword is not known to archy

### CODE REVIEW

File: src/tests/functions/gif/test_gif.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 import pytest
 
-from functions.gif.main import DEFAULT_GIF, UNKNOWN_GIF, gif, gifs
+from functions.gif.main import DEFAULT_GIF, gif, gifs
```
Comment (ZacharyPeloquin) [inline src/tests/functions/gif/test_gif.py:5]: Maybe rename things ? import gif and gifs is very confusing

File: src/tests/functions/gif/test_gif.py [hunk 2]
```diff
@@ -21,15 +21,16 @@ def test_gif_with_param(body):
     assert result == gifs[body["params"][0]]
 
 
-def test_gif_error():
+def test_gif_recherche():
```
Comment (ZacharyPeloquin) [inline src/tests/functions/gif/test_gif.py:24]: function names should be in english

File: src/tests/functions/gif/test_gif.py [hunk 3]
```diff
@@ -21,15 +21,16 @@ def test_gif_with_param(body):
...
     body = {"params": ["JamesDoe"]}
 
     request_mock = MagicMock()
     request_mock.get_json.return_value = body
 
     result = gif(request_mock)
+    url_part = "https://media.tenor.com/"
 
-    assert result == UNKNOWN_GIF
+    assert url_part in result
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:33]: You are not testing anything here ?

File: src/functions/gif/main.py [hunk 4]
```diff
@@ -15,9 +24,15 @@ def gif(request):
     request_json = request.get_json(silent=True)
     if request_json:
         params = request_json.get("params", [""])
+        api_key = os.getenv("API_TOKEN")
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:27]: Change it for `os.environ["API_TOKEN"]`
No need for dotenv

File: src/functions/gif/main.py [hunk 5]
```diff
@@ -1,4 +1,13 @@
+import json
+import os
+
 import functions_framework
+import requests
+from dotenv import load_dotenv
+
+# need nev file with tenor api key
+load_dotenv()
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:9]: You can delete those two lines

File: src/functions/gif/main.py (hunk 4 above)
Comment (ArmandBriere) [reply src/functions/gif/main.py:27]: Rename API_TOKEN to "TENOR_API_TOKEN"
And add this var to the `main.tf` in terraform folder

File: src/functions/gif/main.py [hunk 6]
```diff
@@ -15,9 +24,15 @@ def gif(request):
...
     if request_json:
         params = request_json.get("params", [""])
+        api_key = os.getenv("API_TOKEN")
         if len(params) == 0:
             return DEFAULT_GIF
         if params[0].lower() in gifs:
             return gifs[params[0]]
+        api_request = requests.get(f"https://g.tenor.com/v1/search?q={params[0].lower()}&key={api_key}&limit=1")
+        if api_request.status_code == 200:
+            top_1gifs = json.loads(api_request.content)
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:34]: `top_gif` or `top_one_gif`

File: src/functions/gif/main.py [hunk 7]
```diff
@@ -15,9 +24,15 @@ def gif(request):
...
+            url_gif = top_1gifs["results"][0]["media"][0]["gif"]["url"]
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:35]: `return top_1gifs["results"][0]["media"][0]["gif"]["url"]`

**Review by ArmandBriere — CHANGES_REQUESTED** [review]

File: src/tests/functions/gif/test_gif.py (hunk 3 above)
Comment (Cor-Lapidem) [reply src/tests/functions/gif/test_gif.py:33]: i mean, i just make sure that the api returns a gif url like it's supposed to

Comment (ArmandBriere) [reply src/tests/functions/gif/test_gif.py:33]: :pika:
Tu as changé le code non ? C'etait pas ca avant XD

File: readme.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
+# Archy discord bot
```
Comment (ArmandBriere) [inline readme.md:1]: Delete this file please
It has been renamed to `README.md`

File: src/.firebaserc [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ArmandBriere) [inline src/.firebaserc:1]: Delete this file

File: src/functions/froge/package-lock.json [hunk 10]
```diff
@@ -0,0 +1,531 @@
+{
```
Comment (ArmandBriere) [inline src/functions/froge/package-lock.json:1]: Delete this file

File: src/functions/froge/package.json [hunk 11]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (ArmandBriere) [inline src/functions/froge/package.json:1]: delete this file

File: src/functions/gif/main.py [hunk 12]
```diff
@@ -1,6 +1,9 @@
-from typing import Any, Optional
+"""Cloud function for the !gif command"""
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:1]: Delete the top file comment, we do not need this

File: src/functions/gif/main.py [hunk 13]
```diff
@@ -14,13 +17,33 @@
 @functions_framework.http
 def gif(request):
     """HTTP Cloud Function."""
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:19]: Change the function docstring to something more indicative
"""Return the requested gif or the closest one found with the tenor API."""

Something like this

File: src/functions/gif/main.py [hunk 14]
```diff
@@ -14,13 +17,33 @@
...
-
-    request_json: Optional[Any] = request.get_json(silent=True)
+    request_json = request.get_json(silent=True)
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:20]: Add typing back 
`request_json: Optional[Any] = request.get_json(silent=True)`

File: src/functions/gif/main.py [hunk 15]
```diff
@@ -14,13 +17,33 @@
...
             return DEFAULT_GIF
         if params[0].lower() in gifs:
             return gifs[params[0]]
-        return UNKNOWN_GIF
+        api_request = requests.get(
+            f"https://tenor.googleapis.com/v2/search?q={params[0].lower()}&key={api_key}&client_key=Archy&limit=1"
+        )
+        # https://g.tenor.com/v1/search?q={params[0].lower()}&key={api_key}&limit=1
+
+        # print(api_request.content)
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:34]: Delete dead code

File: src/functions/gif/main.py [hunk 16]
```diff
@@ -14,13 +17,33 @@
...
+
         if len(params) == 0:
             return DEFAULT_GIF
         if params[0].lower() in gifs:
             return gifs[params[0]]
-        return UNKNOWN_GIF
+        api_request = requests.get(
+            f"https://tenor.googleapis.com/v2/search?q={params[0].lower()}&key={api_key}&client_key=Archy&limit=1"
+        )
+        # https://g.tenor.com/v1/search?q={params[0].lower()}&key={api_key}&limit=1
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:32]: delete dead code

File: src/readme.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
+# tox
```
Comment (ArmandBriere) [inline src/readme.md:1]: delete this file

File: src/tests/functions/gif/test_gif.py [hunk 18]
```diff
@@ -1,8 +1,10 @@
-from unittest.mock import MagicMock
+"""Tests for !gif command"""
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:1]: Delete the top comment

File: src/tests/functions/gif/test_gif.py [hunk 19]
```diff
@@ -12,7 +14,9 @@
         {"params": ["confused"]},
     ],
 )
+@patch.dict(os.environ, {"TENOR_API_TOKEN": "{}"})
 def test_gif_with_param(body):
+    """Test known keyword"""
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:19]: Technically, the test name describe what we are testing
We can keed the docstring if you want, we might want to add this to all tests in the future if we keep it

File: src/tests/functions/gif/test_gif.py [hunk 20]
```diff
@@ -21,18 +25,24 @@ def test_gif_with_param(body):
...
+@patch.dict(os.environ, {"TENOR_API_TOKEN": "{}"})
+def test_gif_search():
+    """Test unknown keyword (cover)"""
     body = {"params": ["JamesDoe"]}
 
     request_mock = MagicMock()
     request_mock.get_json.return_value = body
 
     result = gif(request_mock)
+    result = "https://c.tenor.com/JLwgRWNpmYYAAAAC/confusion-chicken.gif"
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:37]: You are hardcoding the solution here, remove this line ?

File: src/tests/functions/gif/test_gif.py [hunk 21]
```diff
@@ -52,3 +64,50 @@ def test_gif_empty_params():
...
+                        498
+                        ],
+                        "size": 7986816
+                    }
+                    }
+                }
+                ]
+            }"""
+
+    result = "https://c.tenor.com/HvtIOMNuwhYAAAAC/wat-what.gif"
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:89]: `result` vs `expected_result` is not the same
The `result` var should be what comes back from the function call

File: src/tests/functions/gif/test_gif.py [hunk 22]
```diff
@@ -52,3 +64,50 @@ def test_gif_empty_params():
...
+    api_status = 200
+
+    assert result == get_gif_from_api(api_status, api_request_json)
+
+
+def test_success_200_bad_json():
+    """Test bad json format api"""
+    api_request_json = """{ }"""
+
+    result = "https://tenor.com/view/im-dead-vanilla-patay-na-ako-dead-nako-patay-ako-gif-22020482"
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:100]: `result` vs `expected_result` is not the same
The `result` var should be what comes back from the function call

File: src/tests/functions/gif/test_gif.py [hunk 23]
```diff
@@ -52,3 +64,50 @@ def test_gif_empty_params():
...
+    api_status = 200
+
+    assert result == get_gif_from_api(api_status, api_request_json)
+
+
+def test_fail_not_200():
+    """Test API request failure"""
+    api_request_json = """{ }"""
+
+    api_status = 503
```
Comment (ArmandBriere) [inline src/tests/functions/gif/test_gif.py:111]: This api status might break the main node if return to it
We want the commands to return code 200 for now

We will add error handling in the main node in the future

File: terraform/function.tf [hunk 24]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (ArmandBriere) [inline terraform/function.tf:1]: delete this file

**Review by ArmandBriere — CHANGES_REQUESTED** [review]

File: src/functions/gif/main.py [hunk 25]
```diff
@@ -14,13 +17,33 @@
...
     """HTTP Cloud Function."""
-
-    request_json: Optional[Any] = request.get_json(silent=True)
+    request_json = request.get_json(silent=True)
     if request_json:
         params = request_json.get("params", [""])
+        api_key = os.environ["TENOR_API_TOKEN"]
+
         if len(params) == 0:
             return DEFAULT_GIF
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:26]: `return (DEFAULT_GIF, 200)`
Adapt test as well

File: src/functions/gif/main.py [hunk 26]
```diff
@@ -14,13 +17,33 @@
...
         if params[0].lower() in gifs:
             return gifs[params[0]]
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:28]: `return (gifs[params[0]], 200)`
Adapt test as well

File: src/functions/gif/main.py [hunk 27]
```diff
@@ -14,13 +17,33 @@
...
-        return UNKNOWN_GIF
+        api_request = requests.get(
+            f"https://tenor.googleapis.com/v2/search?q={params[0].lower()}&key={api_key}&client_key=Archy&limit=1"
+        )
+        # https://g.tenor.com/v1/search?q={params[0].lower()}&key={api_key}&limit=1
+
+        # print(api_request.content)
+
+        return get_gif_from_api(api_request.status_code, api_request.content)
```
Comment (ArmandBriere) [inline src/functions/gif/main.py:36]: `return (return get_gif_from_api(api_request.status_code, api_request.content), 200)`
Adapt test as well

**Review by ArmandBriere — CHANGES_REQUESTED** [review]

**Review by ArmandBriere — APPROVED** [review]

**Pull request merged**